Mocking Singleton does not work even when using @MockitoConfig(convertScopes = true) - Bean produced from factory method

### Describe the bug

We have an issue when using @InjectMock with a `@Singleton `bean and using `@MockitoConfig(convertScopes = true)`
It does not work if our "bean" (sorry if this is not the right term) is created using a factory method / producer method.
The mocking does not seem to work, until I mark the factory method explicitly as `@Produces` - usually you don't have to do this in Quarkus, but to get `@InjectMock` and `@MockitoConfig` to work, it seems to be necessary (to get this mocking to work).

This is a follow on from my comment on here: https://github.com/quarkusio/quarkus/issues/35080



### Expected behavior

Mocking should work and test should run

### Actual behavior

Test does not run

Error is of form as follows:
(please see README.md in enclosed reproducible for full error)
```
...
org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
...
Caused by: java.lang.IllegalStateException: Invalid use of io.quarkus.test.InjectMock - the injected bean does not declare a CDI normal scope but: jakarta.inject.Singleton. Offending field is sampleApiFromFactory of test class class org.acme.SampleUseCaseTest

```

### How to Reproduce?

[sample_app_issue_10022024.zip](https://github.com/quarkusio/quarkus/files/14310941/sample_app_issue_10022024.zip)

Small reproducible attached - please see README.md

### Output of `uname -a` or `ver`

Mac os

### Output of `java -version`

Zulu 21.0.2

### Quarkus version or git rev

3.7.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.3

### Additional information

_No response_

--- a/integration-tests/injectmock/src/main/java/io/quarkus/it/mockbean/SuffixServiceSingletonProducer.java
+++ b/integration-tests/injectmock/src/main/java/io/quarkus/it/mockbean/SuffixServiceSingletonProducer.java
@@ -1,11 +1,10 @@
 package io.quarkus.it.mockbean;
 
-import jakarta.enterprise.inject.Produces;
 import jakarta.inject.Singleton;
 
 public class SuffixServiceSingletonProducer {
 
-    @Produces
+    //@Produces // intentionally commented out to test that auto-producers work with `@InjectMock`
     @Singleton
     public SuffixServiceSingleton dummyService() {
         return new SuffixServiceSingleton();

--- a/integration-tests/injectmock/src/test/java/io/quarkus/it/mockbean/UnusedServiceTest.java
+++ b/integration-tests/injectmock/src/test/java/io/quarkus/it/mockbean/UnusedServiceTest.java
@@ -21,7 +21,7 @@ public class UnusedServiceTest {
     }
 
     @Test
-    public void testNonInjectedUnusedBeanIsNotRemoved() {
+    public void testNonInjectedUnusedBeanIsRemoved() {
         Assertions.assertFalse(Arc.container().instance(OtherUnusedService.class).isAvailable());
     }
 }

--- a/test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/SingletonToApplicationScopedTestBuildChainCustomizerProducer.java
+++ b/test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/SingletonToApplicationScopedTestBuildChainCustomizerProducer.java
@@ -59,7 +59,7 @@ public class SingletonToApplicationScopedTestBuildChainCustomizerProducer implem
                                 continue;
                             }
                             AnnotationValue allowScopeConversionValue = instance.value("convertScopes");
-                            if ((allowScopeConversionValue != null) && allowScopeConversionValue.asBoolean()) {
+                            if (allowScopeConversionValue != null && allowScopeConversionValue.asBoolean()) {
                                 // we need to fetch the type of the bean, so we need to look at the type of the field
                                 mockTypes.add(instance.target().asField().type().name());
                             }
@@ -84,7 +84,15 @@ public class SingletonToApplicationScopedTestBuildChainCustomizerProducer implem
                         context.produce(new AnnotationsTransformerBuildItem(new AnnotationsTransformer() {
                             @Override
                             public boolean appliesTo(AnnotationTarget.Kind kind) {
-                                return (kind == AnnotationTarget.Kind.CLASS) || (kind == AnnotationTarget.Kind.METHOD);
+                                return kind == AnnotationTarget.Kind.CLASS || kind == AnnotationTarget.Kind.METHOD;
+                            }
+
+                            @Override
+                            public int getPriority() {
+                                // annotation transformer registered in `AutoProducerMethodsProcessor` has priority
+                                // of `DEFAULT_PRIORITY - 1` and we need to run _after_ it, otherwise we wouldn't
+                                // recognize an auto-producer (producer without `@Produces`)
+                                return DEFAULT_PRIORITY - 10;
                             }
 
                             @Override
@@ -100,9 +108,8 @@ public class SingletonToApplicationScopedTestBuildChainCustomizerProducer implem
                                     }
                                 } else if (target.kind() == AnnotationTarget.Kind.METHOD) { // CDI producer case
                                     MethodInfo methodInfo = target.asMethod();
-                                    if ((methodInfo.annotation(DotNames.PRODUCES) != null)
-                                            && (Annotations.contains(transformationContext.getAnnotations(),
-                                                    DotNames.SINGLETON)
+                                    if (Annotations.contains(transformationContext.getAnnotations(), DotNames.PRODUCES)
+                                            && (Annotations.contains(transformationContext.getAnnotations(), DotNames.SINGLETON)
                                                     || hasSingletonBeanDefiningAnnotation(transformationContext))) {
                                         DotName returnType = methodInfo.returnType().name();
                                         if (mockTypes.contains(returnType)) {

